### PR TITLE
Scheduler.reschedule() works only by accident

### LINKED ISSUE #6340: Scheduler.reschedule() does not work
Scheduler.reschedule() is called:
1. from Worker.transition_long_running_rescheduled
2. from Worker.transition_executing_rescheduled
3. from stealing.py
4. directly from a bunch of unit tests

If you add "assert False" at the top of the method, only the tests that invoke it directly fail.
Notably, among them there's test_reschedule - in other words, there is nothing at the moment that tests that worker transitions and work stealing are successfully using the method.

### PR DESCRIPTION
**Pull request by crusaderky**
Supersedes #6307
Closes #6340

### CODE REVIEW

**Comment by mrocklin** [comment]
Looks like a valid failing test

**Comment by crusaderky** [comment]
Hilarious. The test passed when the worker was NOT calling Scheduler.reschedule due to mismatched signature. Now that it does, it breaks the state machine. Which has suddenly bumped up the importance and scope of this PR.

This now closes #6340.
I'm parking it for the time being.

File: distributed/tests/test_scheduler.py [hunk 1]
```diff
@@ -1514,30 +1514,33 @@ async def test_log_tasks_during_restart(c, s, a, b):
...
+    assert len(a.state.tasks) == len(b.state.tasks) == 50
 
-    while len(s.tasks) < len(x) + len(futures):
-        await asyncio.sleep(0.001)
+    ys = c.map(slowinc, range(100), key="y", delay=0.1, workers=[a.address])
+    while len(a.state.tasks) != 150:
+        await asyncio.sleep(0.01)
 
-    for future in x:
-        s.reschedule(key=future.key)
```
Comment (crusaderky) [inline distributed/tests/test_scheduler.py:1529]: This test remained green if you removed these lines

File: distributed/tests/test_scheduler.py [hunk 2]
```diff
@@ -3343,12 +3349,11 @@ async def test_worker_heartbeat_after_cancel(c, s, *workers):
 
 @gen_cluster(client=True, nthreads=[("", 1)] * 2)
 async def test_set_restrictions(c, s, a, b):
-
-    f = c.submit(inc, 1, workers=[b.address])
+    f = c.submit(inc, 1, key="f", workers=[b.address])
     await f
     s.set_restrictions(worker={f.key: a.address})
     assert s.tasks[f.key].worker_restrictions == {a.address}
-    s.reschedule(f)
```
Comment (crusaderky) [inline distributed/tests/test_scheduler.py:3351]: This was silently doing nothing (future `f` is not in s.tasks)

File: distributed/tests/test_worker.py [hunk 3]
```diff
@@ -1180,20 +1181,27 @@ def some_name():
     assert result.startswith("some_name")
 
 
-@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)] * 2)
-async def test_reschedule(c, s, a, b):
-    await s.extensions["stealing"].stop()
+@pytest.mark.slow
+@pytest.mark.parametrize("long_running", [False, True])
```
Comment (crusaderky) [inline distributed/tests/test_worker.py:1185]: Added test for `_transition_long_running_rescheduled`

**Review by hendrikmakait — APPROVED** [review]
Code looks good to me, feel free to ignore the comments.

File: distributed/scheduler.py [hunk 4]
```diff
@@ -6568,7 +6568,9 @@ async def get_story(self, keys_or_stimuli: Iterable[str]) -> list[tuple]:
 
     transition_story = story
 
-    def reschedule(self, key=None, worker=None):
+    def reschedule(
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply distributed/scheduler.py:6571]: Well no. A transition to released will automatically kick the task back to waiting.
The functional description of the method is accurate. The "released" bit is an implementation detail.
I'm adding a comment to explain.

[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/scheduler.py:6607]: Makes sense from that perspective, I was thinking that the function doesn't include the rescheduling bit, but in the end you're right, the releasing/descheduling automatically achieves the aim of rescheduling.

File: distributed/tests/test_reschedule.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        if get_worker().address == a_address:
+            raise Reschedule()
+
+    futures = c.map(f, range(4), key=["x1", "x2", "x3", "x4"])
+    futures2 = c.map(slowinc, range(10), delay=0.1, key="clog", workers=[a.address])
+    await wait(futures)
+    assert any(isinstance(ev, RescheduleEvent) for ev in a.state.stimulus_log)
+    assert all(f.key in b.data for f in futures)
+
+
```
Comment (crusaderky) [inline distributed/tests/test_reschedule.py:83]: the two tests below are new

File: distributed/worker_state_machine.py [hunk 6]
```diff
@@ -2201,6 +2196,7 @@ def _transition_released_forgotten(
         ("cancelled", "missing"): _transition_cancelled_released,
         ("cancelled", "waiting"): _transition_cancelled_waiting,
         ("cancelled", "forgotten"): _transition_cancelled_forgotten,
+        ("cancelled", "rescheduled"): _transition_cancelled_released,
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:2199]: Without this test_cancelled_reschedule would cause a RecursionError

**Comment by crusaderky** [comment]
@hendrikmakait there have been substantial changes from your review; could you give it a second pass please?

File: distributed/tests/test_reschedule.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        await asyncio.sleep(0.01)
+
+    # Reschedule the 50 xs that are processing on a
+    for x in xs:
+        if s.tasks[x.key].processing_on is s.workers[a.address]:
+            s.reschedule(x.key, stimulus_id="test")
+
+    # Wait for at least some of the 50 xs that had been scheduled on a to move to b.
+    # This happens because you have 100 ys processing on a and 50 xs processing on b,
+    # so the scheduler will prefer b for the reschduled tasks to obtain more equal
```
Comment (hendrikmakait) [inline distributed/tests/test_reschedule.py:45]:
```suggestion
    # so the scheduler will prefer b for the rescheduled tasks to obtain more equal
```

Comment (hendrikmakait) [reply distributed/tests/test_reschedule.py:45]: I like the comment!

**Review by hendrikmakait — APPROVED** [review]
Good catch with the rescheduling of cancelled tasks!

**Pull request merged**